File: backend/schemas/storyboard.py
"""Pydantic schemas for the Storyboard API surface (Phase 20 — T-251).

These models define the owner and public API contracts for Storyboard launch
keynotes. The strict structured-output payload models the LLM must produce
(``StoryboardPayload`` and friends) live in ``prompts/storyboard.py`` (T-253);
here the generated ``content`` / ``presentation`` payload is carried as an
already-validated JSON object so read endpoints never re-run generation-time
validation.

Privacy invariant: every public-facing model sets ``extra="forbid"`` so adding
any field to an unauthenticated surface is a deliberate, reviewable change and
never a silent ORM passthrough. The public response is allow-list based and must
never expose account email, user id, workspace id, credit data, billing data,
source stage version ids, raw prompts, or previous versions — see
``harness/schemas/storyboard-public-response.schema.json``.
"""

from __future__ import annotations

from datetime import datetime
from typing import Any, Literal
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field

# Lifecycle states for a generated Storyboard version. ``generating`` is the
# in-flight placeholder, ``ready`` is presentable, ``failed`` is a refunded dead
# version, and ``stale`` marks a once-ready version whose source stages were
# refinalised.
StoryboardStatus = Literal["generating", "ready", "failed", "stale"]

# Owner download artifacts. ``html`` and ``pdf`` are rendered packages (T-255);
# ``notes`` / ``demo-script`` / ``appendix`` are markdown documents.
StoryboardDownloadKind = Literal["html", "pdf", "notes", "demo-script", "appendix"]

# Notes can be downloaded as raw markdown or a rendered PDF.
StoryboardNotesFormat = Literal["md", "pdf"]

# Public downloads are a strict subset of owner downloads: the offline HTML
# package is never offered publicly, and every kind here is gated behind the
# matching ``allow_*`` permission.
StoryboardPublicDownloadKind = Literal["pdf", "notes", "demo-script", "appendix"]


class StoryboardSharePermissions(BaseModel):
    """The four independent public-share permission toggles.

    Defaults mirror the column defaults: PDF download on, everything else off.
    """

    allow_pdf_download: bool = True
    allow_notes_download: bool = False
    allow_appendix_download: bool = False
    allow_source_layer: bool = False

    model_config = ConfigDict(extra="forbid")


# ---------------------------------------------------------------------------
# Owner responses
# ---------------------------------------------------------------------------


class StoryboardSummary(BaseModel):
    """Lightweight owner list item — no heavy payload, notes, or source map."""

    id: UUID
    workspace_id: UUID
    version: int
    status: StoryboardStatus
    title: str
    theme: str
    public_share_enabled: bool
    public_share_slug: str | None = None
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(extra="forbid")


class StoryboardDetail(BaseModel):
    """Full owner view of a single Storyboard version.

    Carries the validated presentation payload and source map. The markdown
    artifacts (speaker notes, demo script, appendix) are fetched via the
    download endpoints rather than inlined here to keep the response bounded.
    """

    id: UUID
    workspace_id: UUID
    version: int
    status: StoryboardStatus
    title: str
    theme: str
    content: dict[str, Any]
    source_map: dict[str, Any]
    public_share_enabled: bool
    public_share_slug: str | None = None
    permissions: StoryboardSharePermissions
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(extra="forbid")


class StoryboardPresenterResponse(BaseModel):
    """Owner presenter-mode view: slides plus per-slide speaker notes.

    ``notes`` is the speaker-note map keyed by slide id, lifted from the
    validated payload so presenter mode does not depend on the rendered
    markdown artifact.
    """

    id: UUID
    title: str
    status: StoryboardStatus
    theme: str
    content: dict[str, Any]
    notes: dict[str, Any]
    demo_script_md: str

    model_config = ConfigDict(extra="forbid")


# ---------------------------------------------------------------------------
# Requests
# ---------------------------------------------------------------------------


class StoryboardGenerateRequest(BaseModel):
    """Body for full generation / regeneration (25 credits).

    No fields today — workspace and user are taken from the path and the
    authenticated session — but kept as an explicit model so future generation
    options (e.g. theme hints) are an additive, reviewable change. ``extra`` is
    forbidden so unknown client fields are rejected rather than silently dropped.
    """

    model_config = ConfigDict(extra="forbid")


class StoryboardRegenerateSectionRequest(BaseModel):
    """Body for single-section regeneration (5 credits).

    The section id is a path parameter; the body is reserved for future options
    and currently carries no fields.
    """

    model_config = ConfigDict(extra="forbid")


class StoryboardShareRequest(BaseModel):
    """Enable public sharing and/or update permissions.

    Each permission is optional: ``None`` leaves the current value unchanged (or
    falls back to the default on first enable). The slug is server-generated and
    never client-supplied.
    """

    allow_pdf_download: bool | None = None
    allow_notes_download: bool | None = None
    allow_appendix_download: bool | None = None
    allow_source_layer: bool | None = None

    model_config = ConfigDict(extra="forbid")


class StoryboardShareResponse(BaseModel):
    """Owner-facing result of enabling/rotating a public share."""

    slug: str
    url: str
    enabled: bool
    permissions: StoryboardSharePermissions

    model_config = ConfigDict(extra="forbid")


# ---------------------------------------------------------------------------
# Public response (allow-list, unauthenticated)
# ---------------------------------------------------------------------------


class StoryboardPublicResponse(BaseModel):
    """Allow-list contract for GET /storyboards/public/{slug}.

    Schema-locked to ``harness/schemas/storyboard-public-response.schema.json``.
    ``extra="forbid"`` is the privacy guard: this surface must never carry
    ``user_id``, ``workspace_id``, ``credit_ledger_id``, ``source_stage_version_ids``,
    account email, credit balance, billing history, draft/previous versions, or
    private notes/appendix/source excerpts unless the matching permission is on.
    """

    title: str = Field(min_length=1, max_length=200)
    presentation: dict[str, Any]
    permissions: StoryboardSharePermissions
    downloads: list[StoryboardPublicDownloadKind] = Field(default_factory=list)
    shared_at: datetime

    model_config = ConfigDict(extra="forbid")

File: frontend/src/types/storyboard.ts
// Storyboard types (Phase 20 — T-257).
//
// These mirror the backend Pydantic schemas (backend/schemas/storyboard.py),
// the generated payload contract (backend/prompts/storyboard.py), and the
// harness schemas (harness/schemas/storyboard-payload.schema.json and
// storyboard-public-response.schema.json). Keep them in lockstep: the public
// surface is allow-list based on the backend, and adding a field to a public
// type here is a privacy decision there.

// Lifecycle of a generated Storyboard version. `generating` is the in-flight
// placeholder, `ready` is presentable, `failed` is a refunded dead version,
// and `stale` marks a once-ready version whose source stages were refinalised.
export type StoryboardStatus = "generating" | "ready" | "failed" | "stale"

// The four finalised pipeline artifacts every claim and component maps back to.
export type StoryboardSource = "SPEC" | "PLAN" | "HARNESS" | "TASKS"

// The six fixed, ordered top-level acts of the keynote. Validation and Execution
// Plan are intentionally NOT acts.
export type StoryboardSectionTitle =
  | "Opening Thesis"
  | "Product Vision"
  | "Product Walkthrough"
  | "Technical Architecture"
  | "Trust, Security, Reliability"
  | "Launch Close"

// Slide layout kinds the trusted renderer understands.
export type StoryboardSlideType =
  | "hero"
  | "thesis"
  | "product"
  | "walkthrough"
  | "architecture"
  | "trust"
  | "closing"
  | "appendix_pointer"

// Architecture-reveal planes. Every architecture_reveal diagram layers all eight
// (client through recovery); `group` is an optional structural grouping layer.
export type StoryboardLayerKind =
  | "client"
  | "frontend"
  | "api"
  | "data"
  | "llm"
  | "integrations"
  | "trust"
  | "recovery"
  | "group"

// Owner download artifacts. `html` and `pdf` are rendered packages; the rest are
// markdown documents. The HTML package is owner-only and never offered publicly.
export type StoryboardDownloadKind =
  | "html"
  | "pdf"
  | "notes"
  | "demo-script"
  | "appendix"

// Public downloads are a strict subset of owner downloads — no offline HTML
// package — and each is gated behind the matching `allow_*` permission.
export type StoryboardPublicDownloadKind = "pdf" | "notes" | "demo-script" | "appendix"

// Speaker notes can be downloaded as raw markdown or a rendered PDF.
export type StoryboardNotesFormat = "md" | "pdf"

// ---------------------------------------------------------------------------
// Leaf payload models
// ---------------------------------------------------------------------------

// A bounded citation back to a finalised SPEC / PLAN / HARNESS / TASKS source.
export interface SourceRef {
  source: StoryboardSource
  source_id: string
  excerpt: string
}

// Per-slide presenter guidance. The deck stays sparse; depth lives here.
export interface SpeakerNote {
  slide_id: string
  talk_track: string
  transition: string
  timing_seconds: number
  pause_cue: string
  demo_cue: string
  backup_points: string[]
}

// Structured visual descriptor for a slide. `kind` names an inert layout; extra
// descriptor keys are pure data the renderer escapes and never executes.
export interface StoryboardVisual {
  kind: string
  [key: string]: unknown
}

export interface StoryboardSlide {
  id: string
  type: StoryboardSlideType
  headline: string
  visible_text: string
  visual: StoryboardVisual
  speaker_notes_ref: string
  sources: StoryboardSource[]
}

export interface StoryboardSection {
  id: string
  title: StoryboardSectionTitle
  slides: StoryboardSlide[]
}

export interface StoryboardDiagramLayer {
  id: string
  kind: StoryboardLayerKind
  label: string
  summary: string
  source_refs: SourceRef[]
}

export interface StoryboardDiagram {
  id: string
  type: string
  layers: StoryboardDiagramLayer[]
}

// Visual identity: palette (hex), typography mood, motif, transitions, diagrams.
export interface StoryboardTheme {
  palette: string[]
  typography: string
  motif: string
  transition_style: string
  diagram_style: string
}

// The complete structured keynote payload the LLM produces and the renderers
// consume. `source_map` and the diagram `source_refs` carry bounded excerpts;
// `notes` is keyed by slide id.
export interface StoryboardPayload {
  title: string
  theme: StoryboardTheme
  sections: StoryboardSection[]
  diagrams: StoryboardDiagram[]
  source_map: Record<string, SourceRef[]>
  notes: Record<string, SpeakerNote>
  demo_script_md: string
  technical_appendix_md: string
}

// ---------------------------------------------------------------------------
// Public-share permissions (the four independent owner toggles)
// ---------------------------------------------------------------------------

export interface StoryboardSharePermissions {
  allow_pdf_download: boolean
  allow_notes_download: boolean
  allow_appendix_download: boolean
  allow_source_layer: boolean
}

// ---------------------------------------------------------------------------
// Owner responses
// ---------------------------------------------------------------------------

// Lightweight owner list item. `theme` here is the display label (motif), not
// the structured StoryboardTheme object that lives inside the payload.
export interface StoryboardSummary {
  id: string
  workspace_id: string
  version: number
  status: StoryboardStatus
  title: string
  theme: string
  public_share_enabled: boolean
  public_share_slug: string | null
  created_at: string
  updated_at: string
}

// Full owner view of one Storyboard version. `content` is the validated payload
// (its `demo_script_md` / `technical_appendix_md` live there, not at top level).
// The rendered `speaker_notes_md` and the immutable `source_stage_version_ids`
// source pin are owner-only and fetched on demand (presenter / download
// endpoints) rather than always inlined, so they are optional here.
// `source_stage_version_ids` is never exposed on the public surface.
export interface StoryboardDetail {
  id: string
  workspace_id: string
  version: number
  status: StoryboardStatus
  title: string
  theme: string
  content: StoryboardPayload
  source_map: Record<string, SourceRef[]>
  permissions: StoryboardSharePermissions
  public_share_enabled: boolean
  public_share_slug: string | null
  created_at: string
  updated_at: string
  source_stage_version_ids?: Record<string, string>
  speaker_notes_md?: string
}

// Owner presenter-mode view: the deck payload plus the per-slide speaker notes
// map and the demo script, so presenter mode does not depend on the rendered
// markdown artifact.
export interface StoryboardPresenterResponse {
  id: string
  title: string
  status: StoryboardStatus
  theme: string
  content: StoryboardPayload
  notes: Record<string, SpeakerNote>
  demo_script_md: string
}

// ---------------------------------------------------------------------------
// Share requests / responses
// ---------------------------------------------------------------------------

// Enable sharing and/or update permissions. Each field is optional: omitting it
// leaves the current value unchanged (or the default on first enable). The slug
// is server-generated and never client-supplied.
export interface StoryboardShareRequest {
  allow_pdf_download?: boolean
  allow_notes_download?: boolean
  allow_appendix_download?: boolean
  allow_source_layer?: boolean
}

export interface StoryboardShareResponse {
  slug: string
  url: string
  enabled: boolean
  permissions: StoryboardSharePermissions
}

// ---------------------------------------------------------------------------
// Public (unauthenticated, allow-list) response
// ---------------------------------------------------------------------------

// The allow-list contract for GET /storyboards/public/{slug}. It carries the
// permission-filtered presentation payload (private notes / appendix / source
// excerpts are redacted unless their permission is on) and never exposes
// user_id, workspace_id, credit data, billing data, source_stage_version_ids,
// draft/previous/failed versions, or account email.
export interface StoryboardPublicResponse {
  title: string
  presentation: StoryboardPayload
  permissions: StoryboardSharePermissions
  downloads: StoryboardPublicDownloadKind[]
  shared_at: string
}

// A safe not-found sentinel for the public route: an unknown, disabled, or
// rotated slug resolves to null rather than throwing, so the public page can
// render an empty state without an auth redirect or crash.
export type PublicStoryboardLookup = StoryboardPublicResponse | null
